File: MATLAB/getxpvorth.m
function xpvorth = getxpvorth(qk, Xk, PRM)
%XPVORTH Computes orthogonal part of PV-set
%
   nObj = numel(qk);

   % compute the orthogonal component of Xpv
   xpvorth = rand(nObj) * Xk;                                % - project Xk to a random vector
   xpvorth = xpvorth * diag(1 ./ sqrt(sum(xpvorth.^2)));     % - normalize columns
   xpvorth = xpvorth * PRM;                                  % - orthogonalize to global component space
   xpvorth = diag(sqrt(qk ./ sum(xpvorth.^2, 2))) * xpvorth; % - rescale rows
end

File: R/R/getxpvorth.R
#' Generates the orthogonal part for Xpv
#' @param q.k
#' vector with orthogonal distances for cross-validation set for segment k
#' @param X.k
#' matrix with local validation set for segment k
#' @param PRM
#' projecton matrix for orthogonalization of residuals
#'
#' @return
#' A matrix with orthogonal part for Xpv
#'
#' @importFrom stats runif
#'
#' @export
getxpvorth <- function(q.k, X.k, PRM) {

   nobj <- length(q.k)

   # compute the orthogonal component of Xpv
   Xpv.orth <- matrix(runif(nobj^2, -1, 1), nobj, nobj) %*% X.k           # - project Xk to a random vector
   Xpv.orth <- Xpv.orth %*% diag(1 / sqrt(colSums(Xpv.orth^2)))                # - normalize columns
   Xpv.orth <- Xpv.orth %*% PRM                                                # - orthogonalize to global component space
   Xpv.orth <- diag(sqrt(q.k / rowSums(Xpv.orth^2)), nobj, nobj) %*% Xpv.orth  # - rescale rows

   return (Xpv.orth)
}
